Add support for workspace-wide version setting and bumping

poliorcetics opened the pull request:
This PR adds support for the `workspace.package.version` added in Cargo 1.64.0.

I tried to make very atomic commits to make it easier to review since it needed
quite some refactoring, I advise to review commit by commit else it will be quite
heavy to read.

Closes #752

epage commented on tests/cmd/set-version/upgrade_compatible_dependency.toml:
```diff
@@ -4,7 +4,7 @@ status = "success"
 stderr = """
     Upgraded cargo-list-test-fixture-dependency from 0.4.3 to 0.4.5
-Updated dependency cargo-list-test-fixture from 0.4.3 to 0.4.5
+Updated dependency cargo-list-test-fixture-dependency in cargo-list-test-fixture from 0.4.3 to 0.4.5
```
Not a fan of this message change as it puts the focus on information preceding it.

I'd just do this as "Updated dependency in cargo-list-test-fixture from 0.4.3 to 0.4.5"

epage commented on README.md:
```diff
@@ -211,6 +211,7 @@ OPTIONS:
     -p, --package <PKGID>         Package id of the crate to change the version of
     -V, --version                 Print version information
         --workspace               Modify all packages in the workspace
+        --workspace-version       Modify the version set in [workspace.package] (Cargo 1.64.0+)
```
I would rather we not do this.  If the user specifies a package that is in the workspace, we then update this field.

In the future, let's discuss design things like this in the issue before moving too far on the PR

poliorcetics replied:
I did it like that to allow the conflict with `exclude` (since version inheritance can have consequences on even excluded crate), but I can merge with `--workspace` or `--package` where the package inherits the workspace version, see this https://github.com/killercup/cargo-edit/pull/814#issuecomment-1262187887

epage commented on src/bin/set-version/set_version.rs:
```diff
@@ -90,6 +106,7 @@ fn exec(args: VersionArgs) -> CargoResult<()> {
         all,
         dry_run,
         workspace,
+        workspace_version,
```
imo I'm not a fan of commits that make no-op user visible changes like this as seeing the two sides of the change are important.

(overall though, I really appreciate commit breakdown!)

poliorcetics replied:
Merged with the commit that came after that added the functionnality

epage commented on src/bin/set-version/set_version.rs:
```diff
@@ -121,30 +121,85 @@ fn exec(args: VersionArgs) -> CargoResult<()> {
+    let crate_root =
+        dunce::canonicalize(workspace_manifest.path.parent().expect("at least a parent"))?;
+
+    if let Some(ws_deps) = workspace_manifest.get_workspace_dependencies_mut() {
+        update_filtered_dependencies(
+            ws_deps,
```
I'm not too confident in this updating the correct dependencies

poliorcetics replied:
Adding a quick test with an unrelated dependency to the same version, you're correct, thanks for catching it, I'll fix

poliorcetics replied:
Done, this was easier to fix than expected

epage reviewed (commented):
Thanks for taking care of this!

poliorcetics commented:
@epage so the wanted behavior is to update the workspace-wide version if one of the two is true:

1) `--workspace` was given
2) `--package <pkg>` was given where `pkg`'s version is inherited from the workspace

Is that correct ?

poliorcetics commented:
All fixed, there is only the one question about `--workspace-version` merging with `--workspace` (and loosing the conflict with `--exclude`) or not

epage commented:
Sorry for the delay; got caught up in clap v4 release stuff.

> @epage so the wanted behavior is to update the workspace-wide version if one of the two is true:
> 
>     1. `--workspace` was given
> 
>     2. `--package <pkg>` was given where `pkg`'s version is inherited from the workspace
> 
> 
> Is that correct ?

> All fixed, there is only the one question about `--workspace-version` merging with `--workspace` (and loosing the conflict with `--exclude`) or not

A simpler take is that if a package is specified that uses `version.workspace = true` then we automatically update it. 

We already have command-line flag processing separated from iterating over workspace members, so we can just keep that.

poliorcetics commented:
> Sorry for the delay; got caught up in clap v4 release stuff.

No worries, you don't owe me any of your free time 😄 

----

I fixed a bug where workspace declared in the same `Cargo.toml` as a regular crate would update all the `workspace.dependencies`, regardless of if they were concerned by the update and addressed your comments.

I hope there aren't more bugs I missed, I think I tested pretty much everything now 😅

epage commented:
FYI there are some things that aren't quite working out for me with this PR.  I think part of it is the base you are working off of.  I'm trying to get some changes made that will make adding workspace inheritance more comprehensible.

epage commented:
Alright, things are in better shape.

My assumption is that we will partition the selected manifests into two groups: those inheriting and those that aren't.  When we call `update_dependents` for the inheriting case, we pass in a `&[PathBuf]` of all of the dep crate roots that use workspace dependencies.  We likely don't need more logic split out into new functions.  Forcing code sharing when it doesn't quite work can make the code harder to follow than easier.

poliorcetics commented:
So I can close this ? Seems like you added the code to support it ?

poliorcetics commented:
Rebased, it was indeed much much simpler

epage commented on src/bin/set-version/set_version.rs:
```diff
@@ -132,13 +132,37 @@ fn exec(args: VersionArgs) -> CargoResult<()> {
         let next = target.bump(current, metadata.as_deref())?;
         if let Some(next) = next {
             {
                 let mut manifest = LocalManifest::try_new(Path::new(&package.manifest_path))?;
-                manifest.set_package_version(&next);
+                if !manifest.version_is_inherited() {
+                    manifest.set_package_version(&next);
+                }
 
                 shell_status(
                     "Upgrading",
```
I feel like the `if` check should be moved to this wider block.  We could then provide a tailored `shell_status` to for when the upgrade happened via the workspace.version.

We should also have a comment that is the companion to the workspace update that we still need to run in the inherit mode to update dependents

poliorcetics replied:
I'm not sure I fully understood what you meant, is my last version ok ?

epage reviewed (commented):
Wow, even cleaner than I expected!  Good job on finding such a simple solution!  I hadn't thought to take advantage of the existing `update_dependents`

epage commented:
Thanks!
